### ISSUE TITLE
get_progeny Error: no applicable method for 'rename' applied to an object of class "list"

### ISSUE THREAD

**Issue by bbowles1** [saezlab/OmnipathR#119]
I am attempting to run a simple get_progeny command as shown:

```r
net <- decoupleR::get_progeny(organism = 'mouse', top = 500)
```

The output below is shown:
```
[2024-10-19 21:25:49] [SUCCESS] [OmnipathR] Loaded 700239 annotation records from cache.
[2024-10-19 21:25:50] [SUCCESS] [OmnipathR] UniProt (rest.uniprot.org): downloaded 2 records
Error in UseMethod("rename") : 
  no applicable method for 'rename' applied to an object of class "list"
```

I am using decoupleR_2.10.0 and R v4.4.1.

**Comment by deeenes** [comment]
Hi All,

Apologies for getting back to you so late in this issue. First a TLDR, then a little more details.

*TLDR:* Probably I fixed this bug, I appreciate if you test it with the current head of either the devel or master branch. To do that, install the package from GitHub, set up a temporary, empty cache:

```
library(remotes)
remotes::install_github('saezlab/OmnipathR@devel')
```

In new session:

```
library(OmnipathR)
library(decoupler)

tmpcache <- tempdir()
omnipath_set_console_loglevel('trace')
omnipath_set_cachedir(tmpcache)

net <- get_progeny(organism = 'mouse', top = 500)

# if it fails, issue a curl debug:
options(omnipathr.curl_verbose = TRUE)
net <- get_progeny(organism = 'mouse', top = 500)

# and share the log please:
omnipath_log()
# or
omnipath_logfile()
```

*Disclaimer: this is a very fresh change, I tested a number of affected functions, but the CI tests will run only tomorrow.*

As we see from the logs above, all these errors happened due to a failed download from UniProt. This error was difficult to catch, because it happens inconsistently. I found that a request issued to rest.uniprot.org, with typical default TCP/IP and HTTP parameters, results an extremely slow download (<1 Kb/s), and in about 50-80% of the cases, no packet comes from the server for longer than 60 s (!), and at this point most http libraries exit with a variety of errors pointing to truncated transfer (not always the same error, in curl/OpenSSL with HTTP 1.1 I get an SSL error about unexpected EOF, while with HTTP 2 I get "HTTP/2 stream 1 was not closed cleanly").

I configured the keep-alive TCP parameters, sending by default 13 keep-alive packets, one in every 10 seconds, enabling max 130 seconds idle. Interestingly, applying these parameters not only solved the issue, but speeded up the transfer by a factor of 10 (7-8 Kb/s): my example request that took 210 s (!) before, now took only 16 s (still extremely slow, but at least usable). I published these changes in [OmnipathR v3.15.3](https://github.com/saezlab/OmnipathR/releases/tag/v3.15.3) (current devel & master branches).

In this process, I replaced in the OmnipathR package the http backend used by almost all of its downloading functions. I also added more info about downloads in the log, and added an option to capture libcurl debug messages in the log. This means, even if we still experience issues in the future, it will be easy to collect comprehensive information.

**Comment by deeenes** [comment]
Since my last update, I rewrote another major part of the download infrastructure of the package. I also fixed a number of other minor regressions which made some particular downloads fail. I believe the original issues have been fixed, I'm closing this one, and please open new ones if you still experience any erroneous behaviour.

**Comment by TS83** [comment]
I still get an error:
[2025-02-19 12:06:21] [ERROR]   [OmnipathR] Failed to download `https://rest.uniprot.org/uniprotkb/stream?format=tsv&fields=id,accession&query=organism_id:9606%20AND%20reviewed:true` (attempt 3/3); error: embedded nul in string: '\037\x8b\b\0\0\0\0\0\0\xff\\\xbd͖\xab<\xd0F7\016W\x91K0\xe2\177(~\xcd1`\016\xe0>vO\xb22\xc80\031de\x92\xbbO=\xea\xefm6\x99u\xef%d!\xa4RU\xa9T\xea\xfe\xaf\xff\xe7\xff\xfe\177\xff\xd7\xe5\177\xff?\xff\x8f\xff\xa5ӟQ\xb3\xba\xb6\xf8\xdf\xee\xaf\xd9/\xff\x8b\xbf\xf9[Y\xbc\xe3&\x8b\xd6\xf1\xb9\xc7\xc0u\xfa\xc7\xf5\xb7h~\036{\003\xdcdu6\xe4Q\xd3&\xcd\035\xf8\xe1\xf6\xf4o\036-u\xfb\001}\xc5\xdb\xd6eQ3n\xcf\xf5\xc4q}\033\x8e\177E\xb4\xc7Η'\xfe\xba\xb9h\xab\xe7\024m\xfb\xaa\xf2h\xec\x8f,\a\xea\xe3\xe8\x98cW\001\x8diԹ\x9eU=\xf2\xe8U{>v\xc4Q\xb7\177\016\xf7\x8b\xfaa+\003J\x88\xaah\xee\xda\xf8\xac\xfc\xcft\xb8\xa8\x9e\x8e\xa4>\xd1\xf2\xcfz\xeby\xd4\xe7\x83\177Vg'

**Comment by deeenes** [comment]
Thanks @TS83 for reporting, though this single line doesn't provide enough information. 

To collect and share all details about the error, it would be fantastic if you could send a log file from a clean environment, with curl debug enabled:

```
library(remotes)
remotes::install_github('saezlab/OmnipathR@devel')
remotes::install_github('saezlab/decoupleR@devel')
```

In a new session:

```
library(OmnipathR)
library(decoupleR)

tmpcache <- tempdir()
omnipath_set_console_loglevel('trace')
omnipath_set_cachedir(tmpcache)
options(omnipathr.curl_verbose = TRUE)

net <- get_progeny(organism = 'mouse', top = 500)

# and share the log please:
omnipath_logfile()
```

**Comment by TS83** [comment]
Oh, it works now. Not sure why the error occurred. 
I attached the log file with the error.

[omnipathr-20250219-1123.log](https://github.com/user-attachments/files/18868474/omnipathr-20250219-1123.log)
